### ISSUE TITLE
Dumped headers are buffered when not in a terminal

### ISSUE THREAD

**Issue by connesc** [bagder/curl#324]
I try to use cURL for downloading a large file, while still being able to read the HTTP headers as soon as they are available. From a terminal, I can use the `-D` option (`--dump-header`) like this:
```shell
curl -sSf -D - -o 'largefile.bin' 'http://.../largefile.bin'
```
As expected, dumped headers are instantly printed to stdout. Then, the file is downloaded and written to the disk.

However, if cURL is not controlled by a terminal, the dumped headers are buffered, and not printed until the download completes. For instance:
```shell
curl -sSf -D - -o 'largefile.bin' 'http://.../largefile.bin' | cat
```

The same applies if dumped headers are written to a named pipe (`mkfifo`) instead of stdout. Also, the `-N` flag (`--no-buffer`) does not seem to have any effect here (anyway, buffering is still desired for writing the downloaded file to the disk).

While buffering is obviously better for writing the response body, I think that dumped headers could at least be flushed once written.

Is it possible to get this behavior?

**Comment by bagder** [comment]
`-N, --no-buffer` perhaps?

**Comment by connesc** [comment]
I tried, but no effect.

**Comment by bagder** [comment]
Aha! It isn't used in the header callback... How about this little patch, does it make things better? (assuming you build curl from source)

~~~
diff --git a/src/tool_cb_hdr.c b/src/tool_cb_hdr.c
index af9bb87..a29661e 100644
--- a/src/tool_cb_hdr.c
+++ b/src/tool_cb_hdr.c
@@ -73,10 +73,16 @@ size_t tool_header_cb(void *ptr, size_t size, size_t nmemb, void *userdata)
 
   if(heads->config->headerfile && heads->stream) {
     size_t rc = fwrite(ptr, size, nmemb, heads->stream);
     if(rc != cb)
       return rc;
+    if(config->nobuffer) {
+      /* output buffering disabled */
+      int res = fflush(heads->stream);
+      if(res)
+        return failure;
+    }
   }
 
   /*
    * This callback sets the filename where output shall be written when
    * curl options --remote-name (-O) and --remote-header-name (-J) have

**Comment by connesc** [comment]
Thanks for your quick answer!

I get the following compilation error:
```
tool_cb_hdr.c: In function ‘tool_header_cb’:
tool_cb_hdr.c:78:8: error: ‘config’ undeclared (first use in this function)
     if(config->nobuffer) {
        ^
tool_cb_hdr.c:78:8: note: each undeclared identifier is reported only once for each function it appears in
```

**Comment by bagder** [comment]
Sorry, sloppy of me. Improved version coming soon.

**Comment by connesc** [comment]
No problem, let me know if you want to test other patches.

Also, with this patch and the `-N` option, will buffering be also disabled for the downloaded file? If yes, wouldn't it be better to have two distinct options? Or maybe always disabling buffering for headers (regardless the `-N` option)?

**Comment by bagder** [comment]
This is how the patch was meant to be. But yes, this will enable flushing of both headers and content. I guess doing unconditional flushing of headers could be an idea, but frequent flushing in general lowers performance so I'm not convinced.

~~~
diff --git a/src/tool_cb_hdr.c b/src/tool_cb_hdr.c
index af9bb87..ebfb040 100644
--- a/src/tool_cb_hdr.c
+++ b/src/tool_cb_hdr.c
@@ -73,10 +73,16 @@ size_t tool_header_cb(void *ptr, size_t size, size_t nmemb, void *userdata)
 
   if(heads->config->headerfile && heads->stream) {
     size_t rc = fwrite(ptr, size, nmemb, heads->stream);
     if(rc != cb)
       return rc;
+    if(heads->config->nobuffer) {
+      /* output buffering disabled */
+      int res = fflush(heads->stream);
+      if(res)
+        return failure;
+    }
   }
 
   /*
    * This callback sets the filename where output shall be written when
    * curl options --remote-name (-O) and --remote-header-name (-J) have

**Comment by connesc** [comment]
Thanks for the new patch, it works perfectly now!

However disabling buffering for both headers and content is a bit annoying in my case, since I'm writing a large file to the disk.

Correct me if I'm wrong, but most of the time, HTTP headers do not exceed 16 KB. So, they are probably written in a single chunk. Even if there are more than one chunk, I am not sure they will be enough to have a significant impact on performance. Thus, I guess that unconditionally flushing headers should be acceptable.

Otherwise, would it be possible to configure buffering separately for headers and content?

**Comment by bagder** [comment]
I guess that always flushing the header output wouldn't have any particularly negative impact so I'm prepared to just add that fflush() call unconditionally there...

**Comment by connesc** [comment]
Great! Thanks for the follow up.

Could I expect to see this in the next version? Do you have an idea of the release date?

**Comment by bagder** [comment]
Yes, this is now merged into master and will be part of the next release. Planned release date for 7.44.0 is August 12th.

Thanks for your help on this!

**Comment by connesc** [comment]
Perfect, thanks a lot for your involvement!
